### PR TITLE
Pull request review/approval and comment on code

### LINKED ISSUE #124: Commits and pull files
Before we can start a review, we can implement comment on one line of PRs at first.
I think this includes two sub issues. 

* [ ] ~Commit comment #3748~ see #4898
* [x] Pull files comment #3748

### PR DESCRIPTION
**Pull request by lafriks**
This is merge of my started work on PR approval feature and @lunny comment on code.

Will obsolete #2583 and resolve #124 and #3580 

Screenshot:
attels

Work needs to be done:
* [x] Implement UI for code review and comments
* [x] Implement database changes and needed handlers
* [x] Review display in comment section
* [x] Add tests

**EDIT by @JonasFranzDEV** :
TODOs and other stuff from the merged PR lafriks/gitea#2:

**This comment will summarize all progress and updates.**

## Reviews
Reviews should bundle all code comments created for the initial review. All code comments created for this review are linked to the review model.

## Comment types
`CommentTypeCode`: A comment at a line of a code of a PR
`CommentTypeReview`: Notifies the participants about the review

## TODO
- [x] Render code comments on the right place
- [x] Add support for comments on both sides
- [x] Show code comments in the "conversation" after a review got submitted
- [x] ~~Show single code comment in the "conversation"~~ *(won't be implemented, because it would make the comment list bloated. GitHub only shows single comments in the diff view too)*
- [x] Remove dummy data from "code comment form" (line, path, side)
- [x] Add support for "Add single comments"
- [x] ~~Order/Separate code comments by thread/review~~ *(not implemented, see #4389)*
  - **At the moment only the first comment for a line gets shown in the conversation. No replies!** 
- [x] ~~Blocked by go-gitea/git#116~~
- [x] Edit comments
- [x] Delete comments
- [x] Show actions / notifications for review comments
- [x] ~~Show actions / notifications for review (without comments)~~  *(not implemented, see #4388)*
- [x] Fix some problems with invalidating comments written for the previous code
- Improve UI
  - [x] Show comments only on one side (split view)
  - [x] "Cancel" button functionality
  - [ ] Show add comment button only at PR diff views
  - [x] Show forms only to signed in users
  - [x] Preview tab is broken
  - [x] Add icon is shown at hunk line

## Preview(s)

comments_and_review
review comments
comment at conversation detail

**EDIT END**

### CODE REVIEW

**Comment by sapk** [comment]
Maybe we could try to store it in [git-appraise](https://github.com/google/git-appraise) in place of database ? 
You seem to have work on it in #733 and it iwill also obsolete it.

**Comment by lafriks** [comment]
@sapk that won't be that easy. That was my first idea when starting to work on git-appraise but the problem is how to integrate it with gitea forks & PR's architecture. For now everything will be stored in database and later I will see how to get it all working together also with git-appraise.

**Comment by jonasfranz** [comment]
What is the difference between "Add single comment" and "Add comment"?

**Comment by lunny** [comment]
All the code review buttons are not available.

**Comment by jonasfranz** [comment]
I'm currently trying to contribute to this PR by creating a PR at @lafriks repo. I already did some progress there. Feel free to check it out: https://github.com/lafriks/gitea/pull/2

**Comment by lunny** [comment]
@lafriks @JonasFranzDEV Great efforts. Many people wants this feature on v1.5!

**Comment by jonasfranz** [comment]
@lunny My changes from lafriks/gitea#2 are merged. Feedback is welcome and needed!

**Comment by techknowlogick** [comment]
@JonasFranzDEV can you run `make fmt` so the CI passes?

**Comment by jonasfranz** [comment]
Moved to issue description

**Comment by jonasfranz** [comment]
@techknowlogick Done

**Comment by lunny** [comment]
When I click `Add Single Comment`, it returns 500

```
2018/05/14 21:34:29 [.../repo/pull_review.go:77 CreateCodeComment()] [E] CreateCodeComment: exit status 128 - fatal: This operation must be run in a work tree

2018/05/14 21:34:29 [D] Template: status/500
2018/05/14 21:34:29 http: multiple response.WriteHeader calls
```

**Comment by lunny** [comment]
Both of `Start Review`.

**Comment by jonasfranz** [comment]
@lunny I think that this problem was caused by commenting a file in a subdirectory. This problem should be fixed. If not please send more details about what kind of file is and in which directory the file is.

**Comment by jonasfranz** [comment]
Code comments of reviews will be shown in the conversation now. There are some problems with that. (see last points of [TODO list](https://github.com/go-gitea/gitea/pull/3748#issuecomment-388794096)) Please feel free to improve the design or leave feedback.

code comments in conversation

**Comment by lunny** [comment]
It seems `git blame` cannot run in a bare git repo. @JonasFranzDEV

**Comment by hhenkel** [comment]
@lunny According to this [stackoverflow thread](https://stackoverflow.com/questions/4169556/how-to-git-blame-on-the-remote-side-repository) it seems possible.

**Comment by jonasfranz** [comment]
which git version are you using @lunny

**Comment by lunny** [comment]
see the link @JonasFranzDEV posted, it seems you should use `git blame master -- README.txt`

**Comment by jonasfranz** [comment]
@lunny should be fixed now. (drone picked up old commit)

**Comment by porunov** [comment]
Is there a limit on count of files which are changed to be reviewed? Right now Gogs won't show more than 100 changed files (don't know the limits in Gitea). So, it isn't possible to review very big PRs.

**Comment by lunny** [comment]
@JonasFranzDEV yes. It's fixed. new issue:
image

**Comment by lunny** [comment]
@JonasFranzDEV files conflicts.

**Comment by jonasfranz** [comment]
@lunny Done

**Comment by lunny** [comment]
please resolve conflict.

**Comment by jonasfranz** [comment]
@lunny fixed your new issue (Please flush cache to reload new CSS)

**Comment by lunny** [comment]
@JonasFranzDEV yes, I confirm that.
@lafriks for git-appraise, maybe we can save both database and files after this PR merged.
What's the progress of this PR?

**Comment by lunny** [comment]
@JonasFranzDEV I think git-appraise could be another PR.

**Comment by jonasfranz** [comment]
Demo instance: https://try.h.jonasfranz.software/jonas/gitea/pulls/1

**Comment by techknowlogick** [comment]
@JonasFranzDEV I tried out demo, looks great, but when I click on the `Files` tab of the PR it only shows a blank white page.

**Comment by jonasfranz** [comment]
Should work now

**Comment by techknowlogick** [comment]
@JonasFranzDEV Very nice 👌

File: templates/repo/issue/view_content/comments.tmpl
Comment (krombel) [inline templates/repo/issue/view_content/comments.tmpl]: This hard-coded content does not belong here

Comment (krombel) [reply templates/repo/issue/view_content/comments.tmpl]: Just to mention: This comment was added via "comment" on the diff - I did not press any "review" button.
Just to have an example against your statement "GitHub only shows single comments in the diff view too"
But thanks for your work on this either way. I am really excited to see this merged to master :)

Comment (jonasfranz) [reply templates/repo/issue/view_content/comments.tmpl]: This is some sort of placeholder as I stated somewhere before. It is used for ui prototyping. Currently I am searching a way to render the code efficently.

**Comment by jonasfranz** [comment]
@krombel 
> You can add comments even before the pull request is created - so even before an id got created.
I did not expect that and think that is a bug. Is it one or is that intentional?

I think that you mean that the "+" button is also shown in other diff views. If yes, this is already a point on my TODO List (" Show add comment button only at PR diff views").

For the other points, I added a point on the TODO list. Thank you for your review.

**Comment by jonasfranz** [comment]
@krombel @lafriks I've implemented a "real" code preview for code comments. It saves the patch of the commented file while submitting a comment. If a code comment gets loaded it will parse the patch and render the 4 lines above the commented line.

If you (or somebody else) have a better idea of implementing this feature, please let me know.

Example: https://try.h.jonasfranz.software/jonas/gitea/pulls/1#issuecomment-9

**Comment by lafriks** [comment]
@lunny yes they were but in multiple places already so I removed all and moved these to common place

**Comment by sapk** [comment]
@lafriks doing so will break translation for 1.5.0 release. Isn't possible to do it later in an other PR ?

**Comment by lafriks** [comment]
@sapk Only for 3 strings

**Comment by hhenkel** [comment]
I totally understand that you don't want to break anything for a release - but as this is a long awaited feature by many people (to my understanding) I would like to ask you to reconsider if there is no way to get it into 1.5.0

**Comment by lafriks** [comment]
@hhenkel that would set back Gitea 1.5.0 release by at least month maybe even more. Current plan is to release 1.5.0 and probably in a month or two after that 1.6.0 with this feature

**Comment by lunny** [comment]
I would like this could be an emphasis of v1.6

**Comment by jonasfranz** [comment]
@lunny me too! But I have doubts about that we will get it reviewed until 1.6 release. It might be important that we already start the reviewing progress of the parts which are finished. (CC @go-gitea/maintainers ) That would speed up the progress!

**Comment by jonasfranz** [comment]
I've implemented a new function `CutDiffAroundLine` for creating a diff around the commented line. The diff will be saved as patch in comment. This diff will then be rendered in the conversation list.

**Comment by jonasfranz** [comment]
New updates:
* Code comments can be edited now
* Dashboard / Email notifications for code comments will be sent after review gets submitted
* Dashboard / Email notifications for single code comments will be sent

**Comment by jonasfranz** [comment]
Status is now ready for review/comments

**Comment by lunny** [comment]
When I comment on codes and click `start review` button, it returns 500 error. The logs are below:
```log
[git-module] /Users/lunny/gitea-repositories/ttt/git.git: git -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= show-ref --verify refs/pull/1/head
2018/07/08 15:16:08 [.../repo/pull_review.go:77 CreateCodeComment()] [E] CreateCodeComment: object does not exist [id: refs/pull/1/head, rel_path: ]
2018/07/08 15:16:08 [D] Template: status/500
2018/07/08 15:16:08 http: multiple response.WriteHeader calls
```

**Comment by lunny** [comment]
@JonasFranzDEV No, I didn't.

**Comment by phyr0s** [comment]
When will be available this feature?

**Comment by jonasfranz** [comment]
@videmort The feature is currently in the review progress (needs at least 2 approvals by maintainers). I think that we will integrate it into the 1.6 release but I cannot say when it will be ready for master.

**Comment by phyr0s** [comment]
@JonasFranzDEV  thanks for your quick response. Great job!

**Comment by jonasfranz** [comment]
@lunny Could you retry and re-post the error message. Could you also add the base and head branch name of your repo?

**Comment by lunny** [comment]
The error is the same as before.

**Comment by jonasfranz** [comment]
@lunny Add it nevertheless since I updated the error messages.

**Comment by lunny** [comment]
conflict @JonasFranzDEV

**Comment by jonasfranz** [comment]
@lunny Done. May you please update the error message?

**Comment by jonasfranz** [comment]
I don't understand why CI fails since I regenerated CSS and the patch is identical to my code.

**Comment by lafriks** [comment]
@JonasFranzDEV try npm install to update dependencies to latest versions

**Comment by jonasfranz** [comment]
@lafriks Thank you. That worked for me.

File: models/git_diff.go
Comment (cryptix) [inline models/git_diff.go]: I find the next lines a bit confusing, might want to add a comment.

* Why is LeftIDX multuiplied by -1?
* Why is the LeftIdx comment replaced an the RightIdx one appended?

File: models/issue_comment.go
Comment (cryptix) [inline models/issue_comment.go]: is it okay to ignore this error?

ps: sorry, I'm a bit new to this code base but I couldn't find anything about error logging in the style guide.

File: models/issue_comment.go
Comment (cryptix) [inline models/issue_comment.go]: might be more approachable to make this map a concrete type and explain it?

I guess its `file > line > coments`, right?

File: models/pull.go
Comment (cryptix) [inline models/pull.go]: not sure how much `goto` is used in this codebase but couldn't you put the next goroutine in the `} else {` of this `if err != nil {}` ?

File: modules/templates/helper.go
Comment (cryptix) [inline modules/templates/helper.go]: could do without reflect and get a cleaner outline by using this:


```go
switch key := values[i].(type) {
  case string:
  // fill dict

  case map[string]interface{}:
  // copy to dict
}
```

File: routers/init.go
Comment (cryptix) [inline routers/init.go]: is this related to this PR?

File: models/git_diff.go
Comment (cryptix) [inline models/git_diff.go:254]: I find the next lines a bit confusing, might want to add a comment.

* Why is LeftIDX multuiplied by -1?
* Why is the LeftIdx comment replaced an the RightIdx one appended?

File: models/issue_comment.go
Comment (cryptix) [inline models/issue_comment.go:197]: is it okay to ignore this error?

ps: sorry, I'm a bit new to this code base but I couldn't find anything about error logging in the style guide.

File: models/issue_comment.go
Comment (cryptix) [inline models/issue_comment.go]: might be more approachable to make this map a concrete type and explain it with a godoc comment?

I guess its `file > line > coments`, right?

File: models/pull.go
Comment (cryptix) [inline models/pull.go]: not sure how much `goto` is used in this codebase but couldn't you put the next goroutine in the `} else {` of this `if err != nil {}` ?

File: modules/templates/helper.go
Comment (cryptix) [inline modules/templates/helper.go]: could do without reflect and get a cleaner outline by using this:


```go
switch key := values[i].(type) {
  case string:
  // fill dict

  case map[string]interface{}:
  // copy to dict
}
```

File: routers/init.go
Comment (cryptix) [inline routers/init.go]: is this related to this PR?

**Review by cryptix — COMMENTED** [review]
I'm fairly new to this code base and just did a superficial pass of coding style.

Also only focused on the Go code since JS and CSS really isn't my strong suite.

HTH and really excited to have this in.

File: models/git_diff.go
Comment (jonasfranz) [reply models/git_diff.go:254]: > Why is LeftIdx multiplied by -1?

Because we differentiate between "previous" and "proposed" lines. "Previous" lines are negative and "proposed" lines are positive. This is documented at the comment model.

> Why is the LeftIdx comment replaced an the RightIdx one appended?

Both should be appended. I will fix it.

File: models/pull.go
Comment (jonasfranz) [reply models/pull.go]: `goto` just jumps to `REQUIRED_PROCEDURE` and using an else on every `err != nil` check would result in massive code duplication.

Comment (cryptix) [reply models/pull.go]: I understand your duplication concern. Yet the line after the goto tag checks `if err == nil`, which it won't be, since that triggered the jump. This entanglement warrants a comment, I think.

If you want to run that code in all three cases, I'd put it in a func var like this and call it in all three places.

```go
var queHooks = func() {
 // iterate, preapare and add to que...
}
```

File: models/issue_comment.go
Comment (jonasfranz) [reply models/issue_comment.go:197]: Done, by logging it as warning.

File: models/pull.go
Comment (jonasfranz) [reply models/pull.go]: I re-factored the complete procedure. Is it now better?

File: models/issue_comment.go
Comment (lunny) [inline models/issue_comment.go]:
```
[.../repo/pull_review.go:79 CreateCodeComment()] [E] CreateCodeComment: GetRefCommitID[refs/pull/2/head]: object does not exist [id: refs/pull/2/head, rel_path: ]
```
Error come from this line. We should use `baseGitRepo` but not `headGitRepo` here I think.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (jonasfranz) [reply models/issue_comment.go]: I've changed the discussed code. Could you please make sure that it is fixed now? Your bug only occurs if the PR comes from another repo (e.g. fork).

**Review by lunny — APPROVED** [review]

**Comment by appleboy** [comment]
conflicts @JonasFranzDEV

File: models/issue_comment.go
Comment (techknowlogick) [inline models/issue_comment.go:7]: Please add Gitea copyright to this file

**Comment by jonasfranz** [comment]
@appleboy resolved

File: models/issue_comment.go
Comment (jonasfranz) [reply models/issue_comment.go:7]: Done

**Review by appleboy — APPROVED** [review]

**Comment by adelowo** [comment]
🥇 🥇

**Comment by Caballerog** [comment]
Is in the actual release?

Docker pull in 3, 2, 1 ...

Thanks.

**Comment by chupasaurus** [comment]
@Caballerog this feature will be released in next (1.6) release.

**Comment by ptman** [comment]
@chupasaurus has 1.5.0 been released yet?

**Comment by mkienenb** [comment]
@Caballerog If you are impatient to try it out but don't want to build it from source, you can grab it from here the master release link [1] and install it over your existing binary as described [2] in the "Updating to a new version" section.

[1]
https://dl.gitea.io/gitea/master/

[2]
https://docs.gitea.io/en-us/install-from-binary/

That's what I did.

Thank you everyone else for making this happen!

**Comment by sondr3** [comment]
@chupasaurus Gitea always takes a bit longer than predicted to release, partially I think because they accept too many features each release. This has been discussed before in #2283 and probably lots of other places as well 😃

**Comment by yasuokav** [comment]
Nothing happened when I click the review button. Am I doing anything wrong?

**Comment by lunny** [comment]
@sondr3 I think there are two main reasons. One is what you said, another is all of the maintainers are part-time for the project.

**Comment by sondr3** [comment]
That too, I do not envy you the task of maintaining a project of this size, I have one with a magnitude less users and that was still stressful and took a lot of time. Don’t worry about it, it’s a minor quibble.

**Comment by olekukonko** [comment]
when is the  `ETA` for this functionality?

**Comment by lafriks** [comment]
It is already available for quite some time now. Starting with 1.6.0 version

**Comment by olekukonko** [comment]
Thanks for the prompt response.

**Comment by frankgerhardt** [comment]
Nice feature. I have tried both "Request Change" and "Approve" but really got lost as a reviewer when there were newer commits by the developers. Older comments where shown as out of date but for me as reviewer still valid. I could not see how to track all comments and progress on cleaning them up. Is there any documentation about the workflow? Any blog posts or anything else to understand how this shall be used?

**Pull request merged**
